### ISSUE TITLE
Help needed: Simple WebTransport example with Echo Server

### ISSUE THREAD

**Issue by heppth** [private-octopus/picoquic#1751]
I would like to familiarize myself with picoquic and wanted to write a client for this simple echo server: https://webrtc.internaut.com/wt/.

However, I don't quite understand the [documentation](https://github.com/private-octopus/picoquic/blob/master/doc/pico_webtransport.md) and can't help myself from the baton example.

The plan is to establish a connection, send a test message and receive it again. Unfortunately, the message is not received. 

```
typedef struct wt_ctx_t {
	picoquic_cnx_t* cnx;
	h3zero_callback_ctx_t* h3_ctx;
	uint64_t control_stream_id;
} wt_ctx_t;

int webTransport_test()
{
	int ret = 0;
	struct sockaddr_storage server_address;
	int is_name = 0;
	const char* sni = NULL;
	uint64_t current_time = 0;
	picoquic_quic_t* qclient = NULL;
	wt_ctx_t wt_ctx;

	memset(&wt_ctx, 0, sizeof(wt_ctx_t));

	const char* server_name = "webrtc.internaut.com";
	int server_port = 6161;

	wsa_Init();

	// DNS
	ret = picoquic_get_server_address(server_name, server_port, &server_address, &is_name);
	if (sni == NULL && is_name != 0) {
		sni = server_name;
	}

	// Time
	current_time = picoquic_current_time();

	// Client
	qclient = picoquic_create(
		8,
		NULL, /* Cert */
		NULL, /* Key file */
		"D:\\cacert.pem", /* trust file */
		"h3",
		NULL, /* default_callback_fn */
		NULL, /* default_callback_ctx */
		NULL,
		NULL,
		NULL, /* Reset seed is only for servers */
		current_time,
		NULL, /* Not using simulated time */
		"ticket_store.bin", /* ticket store*/
		NULL, /* Only server need the ticket_encryption_key */
		0 /* Only server need the ticket_encryption_key length */);

	uint32_t preferred_version = 0;
	wt_ctx.cnx = picoquic_create_cnx(qclient, picoquic_null_connection_id, picoquic_null_connection_id,
		(struct sockaddr*)&server_address, current_time,
		preferred_version, sni, "h3", 1);

	wt_ctx.h3_ctx = h3zero_callback_create_context(NULL);

	picowt_set_transport_parameters(wt_ctx.cnx);
	picoquic_set_callback(wt_ctx.cnx, h3zero_callback, wt_ctx.h3_ctx);

	h3zero_stream_ctx_t* stream_ctx = picowt_set_control_stream(wt_ctx.cnx, wt_ctx.h3_ctx);	

	ret = picowt_connect(wt_ctx.cnx, wt_ctx.h3_ctx, stream_ctx, "/echo", webTransport_callback, &wt_ctx);
	if (ret) {
		printf("WebTransport: Error connecting\n");
		exit(-1);
	}

	ret = picoquic_start_client_cnx(wt_ctx.cnx);
	if (ret) {
		printf("WebTransport: Error starting client\n");
		exit(-1);
	}
	ret = picoquic_packet_loop(wt_ctx.cnx, 0, server_address.ss_family, 0, 0, 0, webTransport_loop_callback, &wt_ctx);
}

int webTransport_callback(picoquic_cnx_t* cnx, uint8_t* bytes, size_t length, picohttp_call_back_event_t wt_event, struct st_h3zero_stream_ctx_t* stream_ctx, void* path_app_ctx)
{
	wt_ctx_t* wt_ctx = (wt_ctx_t*)path_app_ctx;

	switch (wt_event) 
	{
		case picohttp_callback_connecting:
			printf("WebTransport connecting\n");
			wt_ctx->control_stream_id = stream_ctx->stream_id;
			break;
		default:
			printf("WebTransport unhandled: %d\n", wt_event);
			break;
	}
	return 0;
}

int webTransport_loop_callback(picoquic_quic_t* quic, picoquic_packet_loop_cb_enum cb_mode, void* callback_ctx, void* callback_argv)
{
	wt_ctx_t* wt_ctx = (wt_ctx_t*)callback_ctx;

	if (cb_mode == picoquic_packet_loop_ready) {
		printf("Handshake complete\n");		
	}
	else if (cb_mode == picoquic_packet_loop_after_receive) {
		printf("Message received\n");
	}
	else if (cb_mode == picoquic_packet_loop_after_send) {
		printf("Message sent\n");
	}
	else if (cb_mode == picoquic_packet_loop_port_update) {
		printf("Port update\n");

		// Data stream
		h3zero_stream_ctx_t* stream_ctx = picowt_create_local_stream(wt_ctx->cnx, 1, wt_ctx->h3_ctx, wt_ctx->control_stream_id);
		if (!stream_ctx)
		{
			printf("Error creating stream\n");
			return -1;
		}

		stream_ctx->path_callback = webTransport_path_callback;

		// Send test message		
		int ret = picoquic_add_to_stream(stream_ctx->cnx, stream_ctx->stream_id, (uint8_t*)TEST_MESSAGE, strlen(TEST_MESSAGE), 1);
		if (ret != 0) {
			printf("Error sending message: %d\n", ret);
			return ret;
		}
		else
		{
			printf("Successfully sent test message\n");
		}
	}
	else {
		printf("Unknown cb_mode: %d\n", cb_mode);
	}
	return 0;
}

int webTransport_path_callback(picoquic_cnx_t* cnx, uint8_t* bytes, size_t length, picohttp_call_back_event_t fin_or_event, struct st_h3zero_stream_ctx_t* stream_ctx, void* path_app_ctx)
{
	printf("WebTransport path callback with length: %d\n", length);
	return 0;
}
```

Output is:

> WebTransport connecting
> Handshake complete
> Port update
> Successfully sent test message

**Comment by huitema** [comment]
Yes, we are piling several layers of protocols so it can be confusing.

I think that you should keep the "webTransport_loop_callback" simple for now. It is meant to handle low level network events, and is not directly synchronized with your web transport connection. You only got the "connecting" callback, but you have not received any other callback indicating that the server is responding. That means it is too soon to try creating streams.

Instead, you should copy and adapt the code in `wt_baton_callback`. You started doing that in `webTransport_callback`, but you only captured the "connecting" event. You should should wait for the "connect accepted" event before  creating your first stream. Then, you will probably receive further events as data is received.

**Comment by heppth** [comment]
Thank you @huitema.

I followed your advice and adapted `wt_baton_callback`. In addition, I am now testing against a Kestrel web server whose debug output I can view. Two problems have occurred.

**Problem 1: Missing :authority header**

> Microsoft.AspNetCore.Connections.ConnectionAbortedException: HTTP/3 stream error (ProtocolError): The :authority and/or :path pseudo-headers are missing.
I solved this problem by adjusting the `h3zero_create_connect_header_frame()` method so that it adds the required header. I will create a pull request if everything works in the end.

**Problem 2: H3_SETTINGS_ERROR**

> Microsoft.AspNetCore.Connections.ConnectionAbortedException: HTTP/3 stream error (SettingsError): HTTP/3 datagrams negotiation mismatch. Currently client has it 'False' and server has it 'True'
> Microsoft.AspNetCore.Server.Kestrel.Transport.Quic[13]
> Stream id "0HN6OG5PQER53:00000000" aborted by application with error code 265 because: "HTTP/3 stream error (SettingsError): HTTP/3 datagrams negotiation mismatch. Currently client has it 'False' and server has it 'True'".

Now I am stuck. By calling picowt_set_transport_parameters and later h3zero_protocol_init, h3 datagram support should be enabled. Where could the error be?

Note: When I get everything working, I'll post my example code close this issue.

**Comment by huitema** [comment]
There may some kind of mismatch for the datagram option negotiation, either at the QUIC level or the H3 level. For picoquic, the HTTP3 "client has it 'False'" if the datagram option is not negotiated at the transport level. The call to picowt_set_transport_parameters should take care of that, but obviously in your case it does not.

Could this be a timing issue? Maybe the call to picowt_set_transport_parameters happens too late, when the transport parameters are already committed.

In the baton sample, you can see a call to:
~~~
            if (ret == 0) {
                ret = picoquic_start_client_cnx(cnx);
            }
~~~
Are you sure that this call only happens after you have set the parameters?

**Comment by heppth** [comment]
Yes, `picowt_set_transport_parameters` is called immediately after creating `cnx`. 
`picoquic_start_client_cnx` is the last call before the packet loop.
I think a timing problem can be ruled out

**Comment by huitema** [comment]
Then we should try to get a log of what happens. Do you generate a "qlog" in your test app? There is example of calling `picoquic_set_qlog(quic, qlog_dir);` in `baton_app.c` -- `qlog_dir` is the path to the directory in which you want the log ("." if you are keep it simple...). You need to include the header `#include <autoqlog.h>`. The call to  `picoquic_set_qlog` should be made after creating the `quic` context and before creating the connection.

**Comment by heppth** [comment]
Many thanks for the quick reply. I have created the following qlog: 
[eeddd2f86dd3888e.client.log](https://github.com/user-attachments/files/17070811/eeddd2f86dd3888e.client.log)

**Comment by huitema** [comment]
You did not wait until the end of the connection, and the log remained in binary form. No problem, I used the "picolog" tool to convert it. I see that picoquic is sending the following transport parameters:
~~~
Event nr: 1
[
  3290,
  "transport",
  "parameters_set",
  {
    "owner": "local",
    "initial_max_stream_data_bidi_local": 2097152,
    "initial_max_data": 1048576,
    "initial_max_streams_bidi": 512,
    "idle_timeout": 30000,
    "max_packet_size": 1440,
    "initial_max_streams_uni": 512,
    "initial_max_stream_data_bidi_remote": 65635,
    "initial_max_stream_data_uni": 65535,
    "active_connection_id_limit": 8,
    "max_ack_delay": 10,
    "handshake_connection_id": "15b8512e2f231753",
    "max_datagram_frame_size": 1536,
    "enable_loss_bit": 1,
    "min_ack_delay": 1000
  }
]
~~~
The "max_datagram_frame_size" is present, as expected. However, it is not recognized by the server implementation, which responds with the following parameters:
~~~
Event nr: 7
[
  75703,
  "transport",
  "parameters_set",
  {
    "owner": "remote",
    "original_connection_id": "eeddd2f86dd3888e",
    "stateless_reset_token": "0b8decfe75b692822a85f6fbb3fcbc5e",
    "max_packet_size": 1472,
    "initial_max_data": 16777216,
    "initial_max_stream_data_bidi_local": 65536,
    "initial_max_stream_data_bidi_remote": 65536,
    "initial_max_stream_data_uni": 65536,
    "initial_max_streams_bidi": 100,
    "initial_max_streams_uni": 10,
    "ack_delay_exponent": 8,
    "max_ack_delay": 26,
    "active_connection_id_limit": 4,
    "handshake_connection_id": "a4159d5983c4e026e0",
    "ff03de1a": "43e8"
  }
]
~~~
No datagram!

I think that something is wrong with the "Kestrel" server. Is it possible that it implements a draft version of the "QUIC datagram" specification instead of the final (RFC) version? That would explain the issue.

**Comment by huitema** [comment]
I checked again the old datagram drafts. The transport parameter code has not changed since the early draft, so a code mismatch is unlikely. Could it be that the "kestrel" server does not implement datagrams? Did you try against the "internaut" server?

**Comment by heppth** [comment]
Based on the log, I assume that Kestrel supports datagrams. I would have to dig deeper into that. What do you mean by “internaut” server?
[kestrel-log.txt](https://github.com/user-attachments/files/17071297/kestrel-log.txt)

**Comment by heppth** [comment]
This seems to be the point where the server breaks the connection:
https://github.com/dotnet/dotnet/blob/main/src/aspnetcore/src/Servers/Kestrel/Core/src/Internal/Http3/Http3Stream.cs#L847

Before my change with the :authority header, the exception in line 837 was thrown.

**Comment by heppth** [comment]
According to Wireshark, “H3 DATAGRAM Draft04” is set at one point. This is probably the problem.
![image](https://github.com/user-attachments/assets/53effe2b-7e6a-43d5-8494-5854e524bfe0)

Here the wireshark dump including keys.
[quic_dump_with_keys.zip](https://github.com/user-attachments/files/17071658/quic_dump_with_keys.zip)

**Comment by huitema** [comment]
I meant the server that you mentioned at the beginning of this thread: https://webrtc.internaut.com/wt/.

**Comment by huitema** [comment]
The setting that you describe is an HTTP3 level setting. It looks obsolete. The HTTP3 datagram service is defined in RFC 9297, which says "An endpoint can indicate to its peer that it is willing to receive HTTP/3 Datagrams by sending the SETTINGS_H3_DATAGRAM (0x33) setting with a value of 1." The RFC was published in August 2022. The value 0xffd77 was defined in the [draft 04](https://datatracker.ietf.org/doc/draft-ietf-masque-h3-datagram/04/) published in October 2021. If we want to interoperate, the "kestrel" server should be upgraded to follow RFC 9297.

That may or may not be the immediate cause of the issue. Datagrams at the HTTP3 level are mapped to QUIC datagrams, defined in RFC 9221. The picoquic stack will not enable support of HTTP3 datagrams if they are not available at the transport layer. The kestrel stack apparently does not support QUIC Datagrams.

Since the kestrel stack is dubious, I suggest that you test interoperability against another server.

**Comment by huitema** [comment]
I started a discussion about this interoperability issue in the Quic developers' slack. The high level summary is that "lots of implementations are sill using old drafts", and that there is no published guideline on how to achieve web transport interoperability between these drafts. But reading the documents, there are probably a few things that could be done in the picoquic implementation.

The first would be to support the fallback to "datagram capsules" when the underlying QUIC connection does not support datagrams. That require some work, but it would allow picoquic to set the "SETTINGS_H3_DATAGRAM" to 1, even when connecting to servers that do not accept QUIC datagrams. That's necessary to solve your problem, but not sufficient.

The Datagram draft went through multiple iterations before converging on RFC 9297. The latest draft versions are compatible with the RFC.

**Comment by huitema** [comment]
@heppth I just merged PR #1757, which brings support for the datagram capsules, and also adds the "authority" field in Extended Connect requests.

The PR also attempts to simplify the coding of new webtransport applications, by moving some of the code in the "baton" demo to the "pico webtransport" library. There is probably more to do there. Lots of the code for setting up a server or a client is boiler plate, and developers should not have to copy that. 

I don't know whether this solves the interop issue with your "kestrel" server. It might, if that server supports the "HTTP3 Datagram" RFC, in addition to the old draft version 04. That old draft was supported by the Chrome browser. I think that Google is planning to upgrade that. Servers like Kestrel will then have to adapt.

**Comment by heppth** [comment]
@huitema First of all, thank you very much for your support. It has helped me a lot to understand QUIC and to continue the evaluation. I am a bit disappointed with the QUIC implementation in asp.net. My goal was to replace the current websocket implementation with a h3 webtransport implementation. After having a closer look at the .net implementation, this will not work at the moment because Microsoft is lagging behind. I am afraid that I will have to postpone this until Microsoft implements the current RFCs.

I am closing the issue, as there are no more open questions on the part of picoquic.
